posts/Swift-Optionals-Are-Monads-Too.md

```diff
@@ -0,0 +1,264 @@
+} ?? "No friends yet." 
+```
+
+This allows us to choose two completely different strings based on whether our `numberOfFriends` variable has a value or not.
+
+One last example of a great usage for `map`. What if we want to read the contents of a url?
+
+```swift
+var webURL: NSURL? = NSURL(string: "http://rouxbe.com/")
+
+var webContents = webURL.map {
+  NSData(contentsOfURL: $0)
+}.map {
+  NSString(data: $0, encoding: NSUTF8StringEncoding)
+}
+
+webContents.map { print($0) }
+```
+
+If I was a bit more mischievous, I could leave the example like that and go on with the post. But the most astute reader who really gets these concepts would point out the mistake in the code sample above. In case you didn't catch it, that's ok. The problem with our code above is that map accepts a closure/function that does not return an `Optional`. `NSData`'s initializer that accepts an `NSURL` is a failable initializer. Since it can return `nil`, our closure can return `nil`. Xcode will complain inside our second call to `map` that `$0` has not been unwrapped. So what do we do? 
+
+### FlatMap to the Rescue
+```swift
+var webURL: NSURL? = NSURL(string: "http://rouxbe.com/")
+
+var webContents = webURL.flatMap {
+  NSData(contentsOfURL: $0)
+}.flatMap {
+  NSString(data: $0, encoding: NSUTF8StringEncoding)
+}
+
+webContents.map { print($0) }
+```
+
+### How do we really apply
+
+Seeing map and flatMap in action should get you started on potential places in code where you can use these concepts or refactor existing logic. But where does apply fit in? When do you ever have an Optional function hanging around that you may want to call with an Optional argument?
+
+Let's dive in with an example. Suppose you have a function with an optional completion handler as one of it's arguments.
+
+```swift
+typealias ParseResponseHandler = (responseData: NSData?) -> [String: AnyObject]
+
+func getJsonFromUrl(urlString: String, 
+    parseResponseHandler: ParseResponseHandler) -> [String: AnyObject]? {
+  let url = NSURL(string: urlString)
+  
+  if let url = url, responseData = NSData(contentsOfURL:url), 
+      parseResponseHandler = parseResponseHandler {
+    return parseResponseHandler(responseData: responseData)
+  } else {
+    return nil
+  }
+}
+```
+
+Notice we have to use if/let syntax to unwrap a slew of optionals before we call the handler with the data. We can simplify this a bit!
+
+```swift
+func getJsonFromUrl(urlString: String, 
+    parseResponseHandler: ParseResponseHandler?) -> [String: AnyObject]? {
+  let url = NSURL(string: urlString)
+
+  return url.flatMap {
+    NSData(contentsOfURL: $0)
+  }.apply(parseResponseHandler)
+}
+```
+
+# Wrapping Up
+
+Hopefully by now you can see how useful functors, applicative functors, and monads can really be. Apple making Optionals work as functors and monads really make writing intentful code a cinch. And just to keep your brain thinking about this crazy topic, Swift's Array type is also a functor and a monad!
```

Comment:
Maybe:

Apple's decision to make Optionals work as functors and monads really encourages writing confident code a cinch. Lastly, just to keep your brain thinking about this crazy topic, Swift's Array type is also a functor and a monad!